### PR TITLE
Editor: Unify the list view sidebar between the post and site editors

### PR DESCRIPTION
**Pull request by youknowriad**
Related #52632 
Similar to #57466 

## What?

This PR continues the work on the great unification between post and site editors. This PR unifies the InserterSidebar component between edit site and edit post.

Worth noting that this PR means the "document overview tab" is also present in the site editor (it was not available there previously).

## Testing instructions

- This component is pretty well covered in terms of e2e test as it represents the main inserter. So there's not really a big need for manual testing.

### CODE REVIEW

File: packages/edit-site/src/components/editor/index.js [hunk 1]
```diff
@@ -59,7 +58,7 @@ import TemplatePartConverter from '../template-part-converter';
 import { useSpecificEditorSettings } from '../block-editor/use-site-editor-settings';
 
 const { BlockRemovalWarningModal } = unlock( blockEditorPrivateApis );
-const { ExperimentalEditorProvider: EditorProvider } =
+const { ExperimentalEditorProvider: EditorProvider, ListViewSidebar } =
```
Comment (ellatrix) [inline packages/edit-site/src/components/editor/index.js]: It was already there, but I guess it's not needed to call it `ExperimentalEditorProvider` if it's a locked API?

Comment (youknowriad) [reply packages/edit-site/src/components/editor/index.js]: I think the reason is that we have a stable and a private one.

**Comment by ellatrix** [comment]
Am I supposed to see the document outline in the site editor? I can't see it when loading a post template

**Comment by youknowriad** [comment]
@ellatrix you mean you see the tab but it's empty right? I'm guessing we may not be targeting the right document, I'll check.

**Comment by ellatrix** [comment]
Screenshot 2024-01-03 at 10 08 46

**Comment by ellatrix** [comment]
It's strange that the word count etc is correct, but there's no outline

**Comment by youknowriad** [comment]
@ellatrix Yeah, it seems the same issue already exist in trunk for the post editor too. Basically, the document outline just retrieves the blocks that are in the root block editor which means it will be wrong when "template preview" is enabled.

I'll see if it's fixable quickly here, otherwise, I'll leave it separate as existing issue.

**Comment by youknowriad** [comment]
@ellatrix The document outline should be fixed now ✅

File: packages/editor/src/components/document-outline/index.js [hunk 2]
```diff
@@ -151,14 +198,14 @@ export const DocumentOutline = ( {
...
-		const { getEditedPostAttribute } = select( editorStore );
+		const { getEditedPostAttribute, getEditorBlocks } =
+			select( editorStore );
 		const { getPostType } = select( coreStore );
 		const postType = getPostType( getEditedPostAttribute( 'type' ) );
 
 		return {
 			title: getEditedPostAttribute( 'title' ),
-			blocks: getBlocks(),
+			blocks: getEditorBlocks(),
```
Comment (ellatrix) [inline packages/editor/src/components/document-outline/index.js]: It does make me wonder if we should include the post title and site title blocks here at some point. Not sure if that makes sense. Perhaps then we should also include article and section boundaries I guess.

**Review by ellatrix — APPROVED** [review]

**Comment by ellatrix** [comment]
Actually there's one more small bug: when you click on the headings in the outline, it doesn't scroll to/select the heading in the editor.

**Comment by ellatrix** [comment]
(And this is working in trunk in the post editor.)

**Comment by youknowriad** [comment]
@ellatrix That's a bit harder to fix. Basically, when using `getEditorBlocks` we retrieve a list of blocks that is different from the one that is rendered if the post/page has not been modified yet. The solution for this would be to have the "block parsing" done in the "core" store and shared between `useEntityBlockEditor` and the `getEditorBlocks` selectors. It's again an issue due to the "transient edits".

**Comment by youknowriad** [comment]
You can notice that as soon as you edit the post/page, clicking the headings work again.

**Comment by youknowriad** [comment]
@ellatrix So it turns out that the issue is even more complex. I initially thought #57500 this would solve it but it won't solve it for the case where "template preview" is enabled (site editor by default or by toggling it in the post editor). The reason is that useBlockSync clones the blocks, so the blocks we have in "core" store are different from the ones we have within the "block editor".  We need to "match" the blocks from the outside. In other words, given a "postId", we need a way to extract the blocks that corresponds to that post id from within the "block-editor" block list and I'm not sure how we can do that.

So for now and to avoid blocking this PR, I'm considering one of these options:

 - Leave it as is: clicking headings don't work unless you make edits.
 - Revert my last commit: document outline is hidden in "template preview" mode (both site and post editors), it's already hidden in post editor when template preview is on on trunk
 - Keep things like trunk: Try to hide the document outline tab entirely from the site editor and "template preview" still broken in post editor.

**Comment by youknowriad** [comment]
After some thinking, I've restored the trunk behavior for now. Only difference between post and site editors is that one lands in "template preview mode" by default and other doesn't.

**Pull request merged**